pkg/controllers/routereflector/ctrl.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	bapi "github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/backend/watchersyncer"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+	log "github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	defaultClusterID                = "224.0.0.0"
+	defaultRouteReflectorMin        = 3
+	defaultRouteReflectorMax        = 10
+	defaultRouteReflectorRatio      = 0.005
+	defaultRouteReflectorRatioMin   = 0.001
+	defaultRouteReflectorRatioMax   = 0.05
+	defaultRouteReflectorLabelKey   = "calico-route-reflector"
+	defaultRouteReflectorLabelValue = ""
+	defaultZoneLabel                = "failure-domain.beta.kubernetes.io/zone"
+)
+
+var notReadyTaints = map[string]bool{
+	"node.kubernetes.io/not-ready":                   false,
+	"node.kubernetes.io/unreachable":                 false,
+	"node.kubernetes.io/out-of-disk":                 false,
+	"node.kubernetes.io/memory-pressure":             false,
+	"node.kubernetes.io/disk-pressure":               false,
+	"node.kubernetes.io/network-unavailable":         false,
+	"node.kubernetes.io/unschedulable":               false,
+	"node.cloudprovider.kubernetes.io/uninitialized": false,
+}
+
+type ctrl struct {
+	updateMutex sync.Mutex
+
+	calicoNodeClient client.NodeInterface
+	k8sNodeClient    k8sNodeClient
+
+	kubeNodeInformer cache.Controller
+	kubeNodes        map[types.UID]*corev1.Node
+
+	calicoNodeSyncer bapi.Syncer
+	calicoNodes      map[string]*apiv3.Node
+
+	bgpPeerSyncer bapi.Syncer
+	bgpPeers      map[string]*apiv3.BGPPeer
+	bgpPeer       bgpPeer
+
+	routeReflectorsUnderOperation map[types.UID]bool
+
+	topology           topologies.Topology
+	datastore          datastores.Datastore
+	incompatibleLabels map[string]*string
+}
+
+func (c *ctrl) Run(stopCh chan struct{}) {
+	if c.datastore.GetType() == apiconfig.EtcdV3 {
+		// Start the Calico node syncer.
+		go c.calicoNodeSyncer.Start()
+	}
+
+	// Start the BGP peersyncer.
+	go c.bgpPeerSyncer.Start()
+
+	// Wait till k8s cache is synced
+	go c.kubeNodeInformer.Run(stopCh)
+	log.Debug("Waiting to sync with Kubernetes API (Nodes)")
+	for !c.kubeNodeInformer.HasSynced() {
+		time.Sleep(100 * time.Millisecond)
+	}
+	log.Debug("Finished syncing with Kubernetes API (Nodes)")
+
+	log.Info("Node controller is now running")
+
+	<-stopCh
+	log.Info("Stopping route reflector controller")
+}
+
+func (c *ctrl) initSyncers(client client.Interface, k8sClientset *kubernetes.Clientset) {
+	if kubeNodes, err := k8sClientset.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{}); err == nil {
```

Comment:
That's a really good question :) So the Kube node watcher waits until the other two resources are in sync here: https://github.com/projectcalico/kube-controllers/pull/514/files#diff-1837ebeba8cdc9148f27b637627e69cbd86cd4afcf30ad4179315b583915eae1R138. The reason is: because it calculates the route reflector topology based on the number of active nodes it must knows about all of the nodes otherwise it should uses with wrong data. Because i haven't found any solution to be sure Kubernetes nodes are in sync, i synced them in the beginning manually, to avoid situations where the operator downscales an existing topology for example, because the node events came from the past and they don't represent the actual status.